### PR TITLE
[flutter_tools] use initially parsed package config for language version, sound mode determination

### LINKED ISSUE #70121: Web: change SDK selection when NullSafetyMode.autodetect is working
From @jonahwilliams 

> [NullSafetyMode.autodetect is when] the user did not pass a null safety flag
> so what we should do is look at the SDK version and entrypoint to determine if it is opted in
> but that is not implemented so it falls back to unsound on web

We didn't fall back to unsound. So in https://github.com/flutter/flutter/pull/70120 I changed it to actually fall back to the unsound SDK.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

There are various parts of the flutter tooling that need to make use of a package config now. parse it once during getBuildInfo and re-use.

Move the null safe language determination here, using the same technique applied on the  web so that the tool can correctly track whether sound or unsound mode is used.

Removes null safety mode autodetect for all modes except test. It is still required here since each individual test file could be opted in or out of null safety.

Additional bonus points: tests that are setting up dummy package_config files can now just create a `PackageConfig` object

Follow up to https://github.com/flutter/flutter/pull/70320
Work towards https://github.com/flutter/flutter/issues/69601
Fixes https://github.com/flutter/flutter/issues/70121

There are still more places where a package config is parsed in a bespoke manner - such as plugins. Fixing that is a bit out of scope.

### CODE REVIEW

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 1]
```diff
@@ -97,38 +93,12 @@ class FlutterDevice {
...
       Artifact platformDillArtifact;
-      List<String> extraFrontEndOptions;
       if (buildInfo.nullSafetyMode == NullSafetyMode.unsound) {
         platformDillArtifact = Artifact.webPlatformKernelDill;
-        extraFrontEndOptions = buildInfo.extraFrontEndOptions;
       } else if (buildInfo.nullSafetyMode == NullSafetyMode.sound) {
         platformDillArtifact = Artifact.webPlatformSoundKernelDill;
-        extraFrontEndOptions =  buildInfo.extraFrontEndOptions;
       } else {
-        final PackageConfig packageConfig = await loadPackageConfigWithLogging(
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/resident_runner.dart:108]: This logic is now the same and shared by all platforms

**Comment by jonahwilliams** [comment]
Need to figure out what to do for `flutter test` - probably skip the null safety mode altogether.

**Review by jmagman — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/resident_runner.dart [hunk 2]
```diff
@@ -96,39 +92,21 @@ class FlutterDevice {
...
-          extraFrontEndOptions =  <String>[
-            ...?buildInfo.extraFrontEndOptions,
-            '--no-sound-null-safety',
-          ];
-          nullSafetyMode = NullSafetyMode.unsound;
+        if (!extraFrontEndOptions.contains('--sound-null-safety')) {
+          extraFrontEndOptions.add('--sound-null-safety');
         }
+      } else {
+        assert(false);
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/resident_runner.dart:109]: This is a bit non-idiomatic. Can this throw a tool exit or StateError instead? Alternately, assert accepts a string as second argument that can give an explanation for the assert.

**Review by zanderso — COMMENTED** [review]
Just a nit from me.

File: packages/flutter_tools/lib/src/resident_runner.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/resident_runner.dart:109]: I count about 14 usages in the flutter tool, all for this sort of enum pattern.

**Pull request merged**